File: gpt-main/gpt-main/server/src/utils/validations.js
const { z } = require('zod');

const trackerUpdateSchema = z.object({
    auditComment: z.string().min(3, "Comment must be at least 3 characters long"),
    uid: z.string().optional(),
    parent_uid: z.string().optional().nullable(),
    vendor: z.string().optional(),
    description: z.string().optional(),
    service_start_date: z.string().or(z.date()).optional().nullable(),
    service_end_date: z.string().or(z.date()).optional().nullable(),
    renewal_month: z.string().or(z.date()).optional().nullable(),
    budget_head: z.string().optional(),
    tower: z.string().optional(),
    contract: z.string().optional().nullable(),
    po_entity: z.string().optional().nullable(),
    allocation_basis: z.string().optional().nullable(),
    initiative_type: z.string().optional(),
    service_type: z.string().optional(),
    fy_budget: z.number().optional(),
    remarks: z.string().optional().nullable(),
    // New PR/PO fields
    pr_number: z.string().optional().nullable(),
    pr_date: z.string().or(z.date()).optional().nullable(),
    pr_amount: z.number().optional().nullable(),
    currency: z.string().optional().nullable(),
    po_number: z.string().optional().nullable(),
    po_date: z.string().or(z.date()).optional().nullable(),
    po_value: z.number().optional().nullable(),
    common_currency: z.string().optional().nullable(), 
    common_currency_value_inr: z.number().optional().nullable(),
});

const monthlyActualUpdateSchema = z.object({
    service_id: z.number().int().positive(),
    entity_id: z.number().int().positive(),
    month_no: z.number().int().min(1).max(12),
    amount: z.number().nonnegative(),
    auditComment: z.string().min(5, "Comment must be at least 5 characters long").optional(),
});

module.exports = {
    trackerUpdateSchema,
    monthlyActualUpdateSchema
};

File: gpt-main/gpt-main/backend/app/utils/validations.py
"""
backend/app/utils/validations.py
Pydantic schemas for request/response validation
"""
from pydantic import BaseModel, EmailStr
from typing import Optional, List

class LoginRequest(BaseModel):
    email: EmailStr
    password: str

class Token(BaseModel):
    access_token: str
    token_type: str

class UserResponse(BaseModel):
    id: int
    name: str
    email: str
    roles: List[str]
    isAdmin: bool

class LoginResponse(BaseModel):
    status: str
    token: str
    data: dict # user object

class BudgetUpdate(BaseModel):
    uid: Optional[str] = None
    parent_uid: Optional[str] = None
    vendor: Optional[str] = None
    service: Optional[str] = None   # mapped to description in some contexts? No, description is separate?
    description: Optional[str] = None
    tower: Optional[str] = None
    budget_head: Optional[str] = None
    contract: Optional[str] = None
    po_entity: Optional[str] = None
    allocation_basis: Optional[str] = None
    initiative_type: Optional[str] = None
    service_type: Optional[str] = None
    currency: Optional[str] = None
    service_start_date: Optional[str] = None # Using string for date input
    service_end_date: Optional[str] = None
    renewal_month: Optional[str] = None
    budget: Optional[float] = None
    actuals: Optional[float] = None
    remarks: Optional[str] = None

class BudgetCreate(BudgetUpdate):
    # Enforce required fields if necessary, currently sticking to Optional to matching existing style
    uid: str
    vendor: str
    description: str

